File: CPP/ByTier/M/1089_M_Minegraphed.cpp
// Problem: CF 1089 M - Minegraphed
// https://codeforces.com/contest/1089/problem/M

/*
 * Purpose: This program designs a 3D game field for the "Minegraphed" game such that the 
 * reachability of labeled positions in the field matches the reachability of vertices in a 
 * given directed graph. 
 * 
 * Algorithm:
 * - The field is constructed as a 3D parallelepiped with dimensions 3n x 3n x 3.
 * - Each vertex of the graph is represented by a labeled cell (1 to n) in the bottom layer.
 * - The connections between vertices (edges in the graph) are encoded as paths in the 3D field:
 *   - A path from vertex i to vertex j is modeled by a sequence of empty cells that allow
 *     movement from i's position to j's position.
 * - A "railway-like" structure is created:
 *   - The top layer has a horizontal track at i*3 + 1 where i is the vertex index.
 *   - The middle layer contains pathways to support transitions between tracks.
 *   - The bottom layer contains all vertex labels.
 * 
 * Time Complexity: O(n^2), due to nested loops scanning and setting up the path connections
 * Space Complexity: O(n^2), for storing the 3D field structure (3 * 3n * 3n)
 */

#include <stdio.h>
#include <algorithm>

using namespace std;
int n;
char s[3][27][27]; // 3D grid storing the 3 layers of the field
int main() {
  scanf("%d", &n);
  // Initialize all cells to '#' (obstacles)  
  for (int i = 0; i < 3; i++)
    for (int j = 0; j < 3 * n; j++)
      fill(s[i][j], s[i][j] + 3 * n, '#');
  
  // Iterate for each vertex in the graph
  for (int i = 0; i < n; i++) {
    // Clear horizontal lines in top and bottom layers for this vertex
    for (int j = 0; j < 3 * n; j++) {
      s[0][i * 3][j] = s[2][j][i * 3] = '.';
    }
    
    // Clear cells in the middle layer to form a path from the top to the bottom for this vertex  
    s[1][i * 3][i * 3 + 1] = s[1][i * 3 + 1][i * 3 + 1] = 
        s[2][i * 3 + 1][i * 3 + 1] = '.';
    
    // For each vertex j, determine if there is an edge from i to j
    for (int j = 0, ti; j < n; j++) {
      scanf("%d", &ti);
      if (ti) {
        // If there's an edge, make a path in the middle layer to connect to this vertex j
        for (int k = 0; k < 3; k++)
          s[k][i * 3 + 1][j * 3 + 1] = '.';
      }
    }
    
    // Place a label (digit) in the bottom layer for vertex i
    s[0][i * 3][n * 3 - 1] = i + '1';
  }
  
  // Output the dimensions of the field
  printf("%d %d %d\n", n * 3, n * 3, 3);
  
  // Output each layer of the field
  for (int i = 0; i < 3; i++) {
    for (int j = 0; j < 3 * n; j++) {
      for (int k = 0; k < 3 * n; k++) {
        putchar(s[i][j][k]); // Print the character at this cell
      }
      putchar('\n');
    }
    if (i != 2)
      putchar('\n'); // Separate layers with a blank line (except the last)
  }
  
  return 0;
}


// https://github.com/VaHiX/CodeForces/

File: Python/ByRound/1089/1089_M_Minegraphed.py
# Problem: CF 1089 M - Minegraphed
# https://codeforces.com/contest/1089/problem/M

"""
Code Purpose:
This code designs a 3D game field that models the reachability of a given directed graph.
It constructs a 3D grid where each vertex of the graph corresponds to a labeled position in the grid.
The movement rules in the game ensure that valid moves in the grid correspond exactly to edges in the graph.
The solution uses a specific layout construction to achieve this mapping.

Algorithms/Techniques:
- Grid construction with specific layout for each vertex
- Direct mapping of graph edges to grid paths using strategic obstacle placement
- Layered 3D grid representation with top-bottom, north-south, and west-east dimensions

Time Complexity: O(n^2) where n is the number of vertices
Space Complexity: O(n^2) for storing the grid and graph representation

The algorithm works by:
1. Creating a 3D grid with dimensions 3*n x 3*n x 3
2. For each vertex i, placing it in the bottom layer with coordinates (3*i+1, 3*i+1)
3. For each edge from i to j, creating paths between these vertices using obstacles
4. Using a specific pattern of obstacles and empty cells to ensure proper movement behavior
"""

import os
import sys

import random
from io import BytesIO, IOBase
import math

input = lambda: sys.stdin.readline().strip()

inf = math.inf

def I():
    return input()

def II():
    return int(input())

def MII():
    return map(int, input().split())

def LI():
    return input().split()

def LII():
    return list(map(int, input().split()))

def LFI():
    return list(map(float, input().split()))

def GMI():
    return map(lambda x: int(x) - 1, input().split())

def LGMI():
    return list(map(lambda x: int(x) - 1, input().split()))

BUFSIZE = 4096

fmax = lambda x, y: x if x > y else y
fmin = lambda x, y: x if x < y else y

class lst_lst:
    def __init__(self, n) -> None:
        self.n = n
        self.pre = []
        self.cur = []
        self.notest = [-1] * (n + 1)

    def append(self, i, j):
        self.pre.append(self.notest[i])
        self.notest[i] = len(self.cur)
        self.cur.append(j)

    def iterate(self, i):
        tmp = self.notest[i]
        while tmp != -1:
            yield self.cur[tmp]
            tmp = self.pre[tmp]


n = II()
grid = [LII() for _ in range(n)]

ans = [[[-1] * (3 * n) for _ in range(3 * n)] for _ in range(3)]

for i in range(n):
    idx = 3 * i + 1

    # Set boundaries of the grid (creating empty spaces)
    for j in range(3 * n):
        ans[0][idx][j] = 0  # Top layer
        ans[2][j][idx] = 0  # Bottom layer
    ans[2][-1][idx] = i + 1  # Label position on bottom layer

    # Create the position marker with obstacles to control movement
    ans[1][idx - 1][idx] = 0  # Empty space
    ans[1][idx - 1][idx - 1] = 0  # Empty space
    ans[0][idx - 1][idx - 1] = 0  # Empty space

    # Set connections based on graph edges
    for j in range(n):
        if grid[i][j]:
            jdx = 3 * j + 1
            ans[0][idx - 1][jdx] = 0  # Empty space for connection
            ans[1][idx - 1][jdx] = 0  # Empty space for connection
            ans[2][idx - 1][jdx] = 0  # Empty space for connection

print(3 * n, 3 * n, 3)
for i in range(3):
    if i:
        print()
    print(
        "\n".join("".join(str(x) if x > 0 else "#."[x + 1] for x in y) for y in ans[i])
    )


# https://github.com/VaHiX/CodeForces/